src/Workspaces/Core/Portable/Execution/Serialization/Checksum.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Execution
+{
+    /// <summary>
+    /// Checksum of data
+    /// </summary>
+    internal sealed partial class Checksum : IObjectWritable, IEquatable<Checksum>
+    {
+        private readonly ImmutableArray<byte> _checkSum;
+        private int? _lazyHash;
+
+        public Checksum(byte[] checksum) :
+            this(ImmutableArray.Create(checksum))
+        {
+        }
+
+        public Checksum(ImmutableArray<byte> checksum)
+        {
+            _lazyHash = null;
+            _checkSum = checksum;
+        }
+
+        public byte[] ToArray()
+        {
+            // TODO: think a way to make this better
+            return _checkSum.ToArray();
```

Comment:
I don't like this because you're implementing in a way that doesn't seem to provide value (using an ImmutableArray) but whihc comes at a cost of allocations.  Given the amount of checksums we're going to create, and hte amount of reading/writing of them we're going to do, i'd like tehm to be as lightweight as possible.

HAving a private byte[] is fine.